### ISSUE TITLE
pyright 1.1.307+ matches overloads for inline function composition incorrectly.

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#5166]
The following [hydra-zen](https://github.com/mit-ll-responsible-ai/hydra-zen) example reveals a discrepancy in how pyright handles overload matching for inline function composition vs multi-step function composition. The following raises for  pyright 1.1.307-1.1.310.

```python
from hydra_zen import builds  # checkout/install hydra-zen commit cdc8ba6

def foo(x: int) -> str: ...

_x = builds(foo, populate_full_signature=True)
fc = builds(_x, populate_full_signature=True)
reveal_type(fc, expected_text="Type[BuildsWithSig[Type[str], (x: int)]]")

# equivalent to above but with inline composition
fc2 = builds(builds(foo, populate_full_signature=True), populate_full_signature=True)
reveal_type(fc2, expected_text="Type[BuildsWithSig[Type[str], (x: int)]]")  # pyright 1.1.307+ raises
```
```
Type text mismatch; expected "Type[BuildsWithSig[Type[str], (x: int)]]" but received "Type[BuildsWithSig[Type[BuildsWithSig[Type[str], (x: int)]], (x: int)]]"
```

I thought that I came up with a more minimal reproducible example, however the following only raises for pyright 1.1.307-1.1.309:

```python
from typing_extensions import Protocol
from typing_extensions import ParamSpec
from typing import TypeVar, overload, Type, Callable, Union

P = ParamSpec("P")
T = TypeVar("T", covariant=True)
R = TypeVar("R")


class A(Protocol[T, P]):
    def __init__(self, *args: P.args, **kwds: P.kwargs):
        ...

def make_a(x: Callable[P, R]) -> Type[A[R, P]]: ...


@overload
def f(x: Type[A[R, P]]) -> Type[A[R, P]]: ...

@overload
def f(x: Callable[P, R]) -> Type[A[R, P]]:  ...

def f(x: Union[Type[A[R, P]], Callable[P, R]]) -> Type[A[R, P]]: ...


def foo(x: int) -> str:
    ...


x = make_a(foo)
y = f(x)
z = f(make_a(foo))

reveal_type(y, expected_text="Type[A[str, (x: int)]]")
reveal_type(z, expected_text="Type[A[str, (x: int)]]")  # pyright raises 
```
```
Type text mismatch; expected "Type[A[str, (x: int)]]" but received "Type[A[A[str, (x: int)], (x: int)]]"
```

I am unable to come up with a minimal example that continues to raise for 1.1.310 as I am seeing for the hydra-zen example.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.

**Comment by rsokl** [comment]
Sorry for the late follow-up on this (I got married 🎉 !)

I have been updating hydra-zen to fix a bunch of type-checking issues that have manifested over the past few weeks. I've ultimately found that the original code snippet (now modified to use `type` instead of `Type`) fails on pyright 1.1.316:

```python
from hydra_zen import builds  # checkout/install hydra-zen commit cdc8ba6

def foo(x: int) -> str: ...

_x = builds(foo, populate_full_signature=True)
fc = builds(_x, populate_full_signature=True)
reveal_type(fc, expected_text="type[BuildsWithSig[type[str], (x: int)]]")

# equivalent to above but with inline composition
fc2 = builds(builds(foo, populate_full_signature=True), populate_full_signature=True)
reveal_type(fc2, expected_text="type[BuildsWithSig[type[str], (x: int)]]")  # pyright 1.1.316 raises
```

My most up-to-date branch can be checked out and installed with:

```
git clone git@github.com:mit-ll-responsible-ai/hydra-zen.git
git fetch origin fix-pyright
cd hydra-zen
pip install .
```

**Comment by erictraut** [comment]
@rsokl, congrats on your recent marriage! 

If you think that you are seeing a new bug, please open a new issue so it doesn't get lost.
